src/Compilers/Core/Portable/DiagnosticAnalyzer/SourceOrNonSourceFile.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// Represents a source file or a non-source file.
+    /// For source files, <see cref="SourceTree"/> is non-null and <see cref="NonSourceFile"/> is null.
+    /// For non-source files, <see cref="NonSourceFile"/> is non-null and <see cref="SourceTree"/> is null.
+    /// </summary>
+    internal sealed class SourceOrNonSourceFile
```

Comment:
>sealed class SourceOrNonSourceFile [](start = 13, length = 34)

Should this be a `readonly struct`?

Reply:
Agree on the `readonly struct` here. The analyzer driven in particular creates lots of these objects. Have a small worry that this could cause an allocation bump. If it's possible to represent this as a `readonly struct` I think we should consider doing so.